datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamSourceCheckpointResources.java

```diff
@@ -0,0 +1,181 @@
+    private final ErrorLogger _errorLogger;
+
+
+    /**
+     * Construct an instance of DatastreamCheckpointResources
+     * @param datastreamServer Datastream server for which health data is retrieved
+     */
+    public DatastreamSourceCheckpointResources(DatastreamServer datastreamServer) {
+        _store = datastreamServer.getDatastreamStore();
+        _coordinator = datastreamServer.getCoordinator();
+        _errorLogger = new ErrorLogger(LOG, _coordinator.getInstanceName());
+        _properties = datastreamServer.getProperties();
+    }
+
+    /**
+     * Get source checkpoints of the parent datastream for each tasks
+     * @return current checkpoint
+     */
+    @Override
+    public ConnectorHealth get() {
+        String datastreamName = getParentKey();
+        try {
+            Datastream datastream = _store.getDatastream(datastreamName);
+            if (datastream == null) {
+                _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_404_NOT_FOUND,
+                        "Datastream does not exist: " + datastreamName);
+
+            }
+            return buildSourceCheckpoint(datastream);
+        } catch (Exception e) {
+            _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_500_INTERNAL_SERVER_ERROR,
+                    "Get datastream checkpoint failed for datastream: " + datastreamName, e);
+        }
+        // Returning null will automatically trigger a 404 Not Found response
+        return null;
+    }
+
+    /**
+     * Update source checkpoint on a given datastream
+     * @param input desired checkpoint value
+     * @return
+     */
+    @Override
+    public UpdateResponse update(ConnectorHealth input) {
+        String datastreamName = getParentKey();
+        Datastream datastream = _store.getDatastream(datastreamName);
+        if (datastream == null) {
+            _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_404_NOT_FOUND,
+                    "Datastream does not exist: " + datastreamName);
+        }
+        // Currently only KafkaCustomCheckpointProvider supports rewinding the checkpoint
+        if (!isCustomCheckpointing(datastream.getConnectorName())) {
+            _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_501_NOT_IMPLEMENTED,
+                    "This datastream doesn't support checkpoints to be updated.");
+        }
+        // Only paused datastreams can be updated
+        if (!DatastreamStatus.PAUSED.equals(datastream.getStatus())) {
+            _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_405_METHOD_NOT_ALLOWED,
+                    "Datastream must be in PAUSED state before updating checkpoint");
+        }
+        try {
+            // Sample SourceCheckpoint resource {"tasks":["sourceCheckpoint":"1234"]}
+            final long newCheckpoint = Long.parseLong(input.getTasks().get(0).getSourceCheckpoint());
+            getKafkaCustomCheckpointProvider(datastream).rewindTo(newCheckpoint);
```

Comment:
BTW, this call doesn't necessarily update the task's local instance. The task will continue to get the old value through KafkaCustomCheckpointProvider::CustomCheckpointProvider.

You either need to send a signal to the task which could be running on any node. Or make the checkpoint provider implementation to not cache and always read from kafka, this has it's own disadvantages.

The simplest way is, you pause the stream then you update the checkpoint and then resume the stream from the streamkeeper. See what is reasonable.

Reply:
I discussed this with Simon and we agreed that the endpoint will throw a message asking user to pause the data stream if the data stream is not in PAUSED state. 

Sample msg:
"_Datastream must be in PAUSED state before updating checkpoint_"